File: university/2025-2/aeds-2/estruturas-de-dados/Arvore Binaria/C/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct No
{
    int elemento;
    struct No* dir;
    struct No* esq;
}No;

typedef struct Arvore
{
    No* root;
}Arvore;

No* newNo(int n);
Arvore* newArvore();
void inserir(Arvore* a, int n);
No* inserirRec(No* root, int n);
void caminhamento(No* root);
void remover(Arvore* a, int n);
No* removerRec(No* root, int n);
No* menorDir(No* i, No* j);


int main()
{
    int n = 10;
    int i;

    Arvore* a = newArvore();

    inserir(a, 3);
    inserir(a, 7);
    inserir(a, 9);
    inserir(a, 4);
    inserir(a, 2);

    caminhamento(a->root);
    printf("\n");

    remover(a, 4);

    caminhamento(a->root);
    printf("\n");
}

No* newNo(int n)
{
    No* temp = (No*) malloc(sizeof(No));
    temp->esq = NULL;
    temp->dir = NULL;
    temp->elemento = n;

    return temp;
}

Arvore* newArvore()
{
    Arvore* temp = (Arvore*) malloc(sizeof(Arvore));
    temp->root = NULL;

    return temp;
}

void inserir(Arvore* a, int n)
{
    a->root = inserirRec(a->root, n);
}

No* inserirRec(No* root, int n)
{
    if(root == NULL)
        root = newNo(n);
    else if(n < root->elemento)
        root->esq = inserirRec(root->esq, n);
    else if(n > root->elemento)
        root->dir = inserirRec(root->dir, n);
    
    return root;
}

void remover(Arvore* a, int n)
{
    a->root = removerRec(a->root, n);
}

No* removerRec(No* root, int n)
{
    if(root == NULL)
        return NULL;
    
    if(n < root->elemento)
        root->esq = removerRec(root->esq, n);
    else if(n > root->elemento)
        root->dir = removerRec(root->dir, n);
    
    else if(root->dir == NULL)
    {   
        No* del = root;
        root = root->esq;
        free(del);
    }
    else if(root->esq == NULL)
    {   
        No* del = root;
        root = root->dir;
        free(del);
    }
    else
        root->dir = menorDir(root, root->dir);
      
    return root;
}

No* menorDir(No* i, No* j)
{
    if(j->esq != NULL)
    {
        j->esq = menorDir(i, j->esq);
    }
    else
    {
        i->elemento = j->elemento;
        No* temp = j->dir;
        free(j);
        j = temp;
    }
    return j;
}

void caminhamento(No* root)
{
    if(root != NULL)
    {
        printf("%d ", root->elemento);
        caminhamento(root->esq);
        caminhamento(root->dir);
    }
}       

File: university/2025-2/aeds-2/estruturas-de-dados/Arvore Binaria/Java/main.java
public class main
{
    public static void main(String[] args)
    {
        Arvore a = new Arvore();

        a.inserir(1);
        a.inserir(15);
        a.inserir(10);
        a.inserir(3);
        a.inserir(2);
        a.inserir(4);
        a.inserir(7);
        a.inserir(5);

        a.caminhamento();
    }

   
}

class No
{
    public int elemento;
    public No dir;
    public No esq;

    public No()
    {
        this.elemento = 0;
        this.esq = null;
        this.dir = null;
    }

    public No(int n)
    {
        this.elemento = n;
        this.esq = null;
        this.dir = null;
    }
}

class Arvore
{
    public No root;

    public Arvore()
    {
        this.root = null;
    }

    public void inserir(int n)
    {
        root = inserirRec(this.root, n);
    }

    public No inserirRec(No root, int n)
    {
        if(root == null)
            root = new No(n);
        else if(n < root.elemento)
            root.esq = inserirRec(root.esq, n);
        else if(n > root.elemento)
            root.dir = inserirRec(root.dir, n);

        return root;
    }

    public void caminhamento()
    {
        caminhamentoRec(this.root);
    }

    public void caminhamentoRec(No root)
    {
        if(root != null)
        {
            System.out.println(root.elemento);
            caminhamentoRec(root.esq);
            caminhamentoRec(root.dir);
        }
    }
}
